### ISSUE TITLE
Issues on armv7l (Raspberry Pi 3 w/ Rasbian Jessie)

### ISSUE THREAD

**Issue by benjaoming** [learningequality/kolibri-installer-debian#30]
Having python3-cryptography as Recommends now fails on Raspberry Pi -- or I'm not quite sure if that's the problem, because even if I remove it, it remains broken!

It seems that python3-cffi needs to be a newer version according to [this comment](https://pagure.io/freeipa/issue/5786#comment-337506):

> It's a bug in our build of the python-cryptography package. python-cryptography needs a more recent version of python-cffi. The box has only python-cffi-0.8.6-2.el7.x86_64 installed. FreeIPA's COPR has python-cffi 1.4.2-1.el7.centos, but it's not getting installed automatically. python-cryptography doesn't have the necessary version requirement in its spec file:
>
> ```
> # rpm -qR python-cryptography | grep cffi
> python-cffi >= 0.8
> ```
>
> It should be python-cffi >= 1.4.

Can we come up with a work-around?

Here's the error...

```
kolibri@raspberrypi:~$ kolibri shell
Traceback (most recent call last):
  File "/usr/bin/kolibri", line 9, in <module>
    load_entry_point('kolibri==0.9.0', 'console_scripts', 'kolibri')()
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 356, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2476, in load_entry_point
    return ep.load()
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2190, in load
    ['__name__'])
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 28, in <module>
    from . import server  # noqa
  File "/usr/lib/python3/dist-packages/kolibri/utils/server.py", line 8, in <module>
    import requests
  File "/usr/lib/python3/dist-packages/kolibri/dist/requests/__init__.py", line 84, in <module>
    from urllib3.contrib import pyopenssl
  File "/usr/lib/python3/dist-packages/kolibri/dist/urllib3/contrib/pyopenssl.py", line 46, in <module>
    import OpenSSL.SSL
  File "/usr/lib/python3/dist-packages/kolibri/dist/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
  File "/usr/lib/python3/dist-packages/kolibri/dist/OpenSSL/crypto.py", line 12, in <module>
    from cryptography import x509
  File "/usr/lib/python3/dist-packages/kolibri/dist/cext/cp34/linux-armv7l/cryptography/x509/__init__.py", line 9, in <module>
    from cryptography.x509.base import (
  File "/usr/lib/python3/dist-packages/kolibri/dist/cext/cp34/linux-armv7l/cryptography/x509/base.py", line 16, in <module>
    from cryptography.x509.extensions import Extension, ExtensionType
  File "/usr/lib/python3/dist-packages/kolibri/dist/cext/cp34/linux-armv7l/cryptography/x509/extensions.py", line 18, in <module>
    from cryptography.hazmat.primitives import constant_time, serialization
  File "/usr/lib/python3/dist-packages/kolibri/dist/cext/cp34/linux-armv7l/cryptography/hazmat/primitives/constant_time.py", line 9, in <module>
    from cryptography.hazmat.bindings._constant_time import lib
AttributeError: 'module' object has no attribute '_init_cffi_1_0_external_module'
```

 * System: Raspbian Jessie, armv7l
 * python3-cryptography version: 0.6.1-1+deb8u1
 * python3-cffi version: 0.8.6-1

This did NOT work, though:

 * `sudo apt remove python3-cryptography`
 * `sudo apt remove python3-openssl`

**Comment by lyw07** [comment]
We bundled cryptography 2.0.3 into raspberry pi, so I guess it might be because some of the libraries in the system are relatively older than this cryptography version? But I will look into this more

**Comment by benjaoming** [comment]
@lyw07 am trying to get a Raspberry Pi 3 installed with Kolibri at the moment... I found out that I have to upgrade the whole thing because apparently a lot of the CA certificates were expired or something like that... doing a system upgrade of some 300 packages.. so I kind of lost track of what exactly is the cause. Tomorrow, it'll be done upgrading :)

**Comment by benjaoming** [comment]
I will also try `sudo easy_install -U cffi` because it seems there are some issues both with the native `pip` and with the base version of `cffi` (as indicated by the traceback).

**Comment by lyw07** [comment]
I installed kolibri 0.9.0-build2 on Debian jessie and got the same error. Don't have this error on Debian Stretch though. Maybe the issue is because kolibri of `ubuntu xenial` is incompatible in debian jessie.

**Comment by lyw07** [comment]
I think for debian jessie/raspberry pi debian jessie, one workaround is to update `cffi` with `pip` or remove `python3-cffi`. Another one can be to directly install kolibri using `gdebi kolibri.deb` instead of getting kolibri from PPA. One thing I'm still confused is that we have `cffi` bundled inside kolibri and the path of `cffi` is in front of the path of `python3-cffi`. The version of `cffi` bundled inside kolibri is the latest version `1.11.5`, which I think shouldn't cause the error here? So it looks like somehow when running kolibri in debian jessie, it gets `cffi` from the system library somewhere.

**Comment by benjaoming** [comment]
> So it looks like somehow when running kolibri in debian jessie, it gets cffi from the system library somewhere.

Could be that the bundled `kolibri/dist/cext` package is ignored because it doesn't match the system platform?

I'm also curious about which packages depend on `python3-cffi` -- I would think it's not an option to remove it, because it seems quite important. Of course, replacing it with a new system-wide `cffi` obtained via `pip` might work, it's just impossible to build into the .deb.

# Workaround 1: PPA

It's possible to add a newer build of `cffi` to our own PPA and then depend on that. We should play around with the naming, though, such that it only gets installed in cases where it's needed and doesn't shadow out a sufficient `cffi` provided by the platform.

Disadvantage: It's going to make it harder to use a redistributable file.

# Workaround 2: Automated failure from Kolibri source

We could also write a script that outputs the right guidance to the user: "Please run `sudo pip install cffi --upgrade`".

# Workaround 3: Bundle `cffi` and add it conditionally

We could also work out exactly which platforms and under which circumstances we need `cffi` and only add it to the path in these cases. Still have to resolve @lyw07 's comment:

> One thing I'm still confused is that we have cffi bundled inside kolibri and the path of `cffi` is in front of the path of `python3-cffi`

**Comment by lyw07** [comment]
> Could be that the bundled kolibri/dist/cext package is ignored because it doesn't match the system platform?

I think the `cffi` package we bundled inside kolibri is the `kolibri/dist/cext/cp34/linux-armv7l/cffi`.

The package depends on `python3-cffi` is [`python3-cryptography`](https://packages.debian.org/jessie/python3-cryptography). 

I wrote:
> One thing I'm still confused is that we have cffi bundled inside kolibri and the path of cffi is in front of the path of python3-cffi

because we have the `cffi` package for the correct platform bundled inside kolibri (`kolibri/dist/cext/cp34/linux-armv7l/cffi`), and in the PYTHONPATH, this path of the `cffi` package is in front of the path of the system `cffi`. I'm confused that the system `cffi` still gets called even if we have the `cffi` package inside kolibri.


I feel that workaround 2 might be better for us currently?
